### PR TITLE
CLI command to chat with an assistant using rasa webchat

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- add a new CLI command `rasa chat` to run the assistant using SocketIO and open webchat
- added HTML page to host the webchat, keeping everything on localhost

**Example interaction**:

https://user-images.githubusercontent.com/1098412/122185610-83178780-ce8d-11eb-9755-8e99cb922705.mov

If we want to add this, we'd need to update the documentation and I'd suggest recommending using the webchat rather than `rasa shell`.

The HTML surrounding the webchat currently is very basic. We'd at least need some background/design for it. (random idea in the future @TyDunn we could use the background to feature conferences or community events - just a random thought though). We could also use the space to show more state information about the conversation, e.g. grabbing the sidepanel we use in Rasa X:
Bildschirmfoto 2021-06-16 um 10 35 51

### CODE REVIEW

**Comment by tmbo** [comment]
I can add necessary tests / documentation, but we should discuss first if this is the path we want to take.

**Comment by TyDunn** [comment]
@tmbo This super cool! Sorry for the delay on this one. It slipped through the cracks. Do you think we should continue to support `rasa shell` as well? If so, why?

**Comment by tmbo** [comment]
I don't think we should; do you think there is a use case that is specific to a cli-only interaction?

**Comment by TyDunn** [comment]
Hmm. Not really. What about users running Rasa Open Source on headless servers? For example, we are using a GCP VMs for user tests today and won't have access to a browser, so we'll use `rasa shell` (and would not be able to use `rasa chat` if it were available)

**Comment by amn41** [comment]
`rasa shell --debug` is pretty helpful

**Comment by tmbo** [comment]
you'll get the same information though talking through the webchat. the downside would be that it isn't "interleaved" with your interaction in the same way. (something we can fix though with better log messages)

**Comment by tmbo** [comment]
> For example, we are using a GCP VMs for user tests today and won't have access to a browser, so we'll use rasa shell (and would not be able to use rasa chat if it were available)

@TyDunn that is a good point and something you can't get around without some tooling to expose the interface (e.g. ngrok OR `rasa --share` 😉 ). 

I think this is a discussion we should also have with the community as there might be other hidden use cases for `rasa shell`.

**Comment by TyDunn** [comment]
> I think this is a discussion we should also have with the community as there might be other hidden use cases for rasa shell.

Yeah, I was talking to a hero last week, and he said that he heavily uses the `debug` flag with `rasa shell`. It's how he resolves issues. For example, he will use it if he wants to check if the event tracker storing things matches their stories and things like that.  He uses `rasa shell `often, although he is moving more towards using automated tests. He mentioned it would be nice if could get the debug information out of rasa tests because then he wouldn't necessarily need to use `rasa shell` so much

**Comment by tmbo** [comment]
@RasaHQ/enable-squad any plans to include this? otherwise I'd go ahead and close as it is going stale

**Comment by TyDunn** [comment]
I think we need to figure out a plan with `rasa shell` if we were to merge this. We don't have too many more urgent priorities right now, so I will close this

File: rasa/core/run.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ def configure_app(
     syslog_port: Optional[int] = None,
     syslog_protocol: Optional[Text] = None,
     request_timeout: Optional[int] = None,
+    server_listeners: Optional[List[Tuple[Callable, Text]]] = None,
```
Comment (ancalita) [inline rasa/core/run.py:97]: What are these server listeners for? Asking out of ignorance really 😄

File: rasa/core/channels/chat.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      data-new-gr-c-s-loaded="14.1014.0">
+<div class="logged-in-box auth0-box logged-in">
+    <h2>Happy chatting!</h2>
+    <div>You can include the chat on your own website using</div>
+    <div><code><pre>
+    &lt;div id="rasa-chat-widget" data-websocket-url="https://example.com/"&gt;&lt;/div&gt;
+&nbsp;&nbsp;&nbsp;&nbsp;&lt;script src="https://unpkg.com/@rasahq/rasa-chat" type="application/javascript"&gt;&lt;/script&gt;
+    </pre></code></div>
+</div>
+make sure to replace <code>https://example.com/</code> with the URL of your server
```
Comment (ancalita) [inline rasa/core/channels/chat.html:13]: Does this refer to the `data-websocket-url` in L9?

File: rasa/cli/chat.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+def chat(args: argparse.Namespace) -> None:
+    """Chat to the bot using the most recent model."""
+    import rasa.cli.run
+
+    args.connector = "socketio"
+
+    async def after_start_hook_open_chat(_: Sanic, __: AbstractEventLoop) -> None:
+        """Hook to open the browser on server start."""
+        server_url = constants.DEFAULT_SERVER_FORMAT.format("http", args.port)
```
Comment (ancalita) [inline rasa/cli/chat.py:61]: How does the server_url work with the `data-websocket-url` from `chat.html` file?

**Review by ancalita — COMMENTED** [review]
I was just checking any PRs I got assigned to and noticed this via the squad assignment 😄 
Looks good 💯 I only asked a few questions out of curiosity.

A few things to follow-up on though:
- How best do we go about testing this works fine?
- We also need to update docs here with instructions on how to set this up, do you plan adding these to this PR?
- A cli test is failing because it needs an assertion updated to contain the new command.
- There are some code quality checks that need fixing + changelog entry.

File: rasa/core/run.py (hunk 1 above)
Comment (tmbo) [reply rasa/core/run.py:97]: they are a more idiomatic way to register hooks to be triggered after the server starts - e.g. for the chat command we want to open the browser after the server is ready.

File: rasa/core/channels/chat.html (hunk 2 above)
Comment (tmbo) [reply rasa/core/channels/chat.html:13]: yes it does

File: rasa/cli/chat.py (hunk 3 above)
Comment (tmbo) [reply rasa/cli/chat.py:61]: I am not sure what you mean?

**Comment by tmbo** [comment]
@chandrikas not sure what to do with this one, I personally am using it but I am not sure what the status of the webchat UI is. I am checking in with the rest of the product team first

**Pull request closed without merging**